### PR TITLE
Refactor: Configure subnetwork trace

### PR DESCRIPTION
**Pull request by yo1995**
This PR refactors the `Configure subnetwork trace` sample by Vivian.

|previous|current|
|-|-|
|current|demo|

### CODE REVIEW

**Comment by yo1995** [comment]
Please take a look and we can discuss further improvements.

Also I was confused about the coded value display. It seems it should display as its name, but your version display it as raw numeric value. I didn't change that.

**Comment by vquach2404** [comment]
I think the addition of the toolbar and the add condition scene is nice. It may be repetitive to have a list of conditions and list them again but chained. Additionally, users can add as many conditions as they want which makes having two lists concerning. I understand that it helps simplify the logic with the deletion feature, but do you think we can stick to a single list?

Also, for attributes with predefined coded values, my implementation displays the name of the coded values (A, AB, etc). Not sure where the confusion is here.

File: arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceViewController.swift [hunk 1]
```diff
@@ -108,462 +159,201 @@ class ConfigureSubnetworkTraceViewController: UITableViewController {
...
             // Check if attribute domain is a coded value domain.
-            if let domain = attributeComparison.networkAttribute.domain as? AGSCodedValueDomain {
-                // Get the coded value using the the attribute comparison value and attribute data type.
-                let dataType = attributeComparison.networkAttribute.dataType
-                let attributeValue = convertToDataType(otherValue: attributeComparison.value!, dataType: attributeComparison.networkAttribute.dataType)
-                let codedValue = domain.codedValues.first(where: { compare(dataType: dataType, value1: $0.code!, value2: attributeValue!) })
-                let comparisonOperatorString = comparisonsStrings[attributeComparison.comparisonOperator.rawValue]
-                return "'\(attributeComparison.networkAttribute.name)' \(comparisonOperatorString) '\(codedValue!.name)'"
+            if let codedValue = attributeComparison.value as? AGSCodedValue {
+                return "'\(attributeName)' \(comparisonOperator) '\(codedValue.name)'"
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceViewController.swift:227]: Put a break point on this line.

File: arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    
+    // MARK: Actions
+    
+    @IBAction func addConditionBarButtonItemTapped(_ sender: UIBarButtonItem) {
+        if let attribute = selectedAttribute, let comparison = selectedComparison, let value = selectedValue {
+            let convertedValue: Any
+            
+            if let codedValue = value as? AGSCodedValue, attribute.domain is AGSCodedValueDomain {
+                // The value is a coded value.
+                convertedValue = convertToDataType(value: codedValue.code!, dataType: attribute.dataType)
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift:82]: > Also, for attributes with predefined coded values, my implementation displays the name of the coded values (A, AB, etc). Not sure where the confusion is here.

I made changes to the `convertToDataType` function, and it no longer give results other than numbers and booleans. In your original implementation it mostly default to return the coded value type itself, while now it becomes a Swift value type. I didn't scrutinize the original design - is converting coded value needed? If not I can make changes to this line.

Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift:82]: Edit: it seems you explicitly convert a coded value back to its name by finding it from all the values in its domain. I can change to this approach.

Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift:82]: Yes, please change it. Thanks!

Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift:82]: It should be good now!

File: arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    
+    func showValueInputField(completion: @escaping (NSNumber?) -> Void) {
+        let alertController = UIAlertController(title: "Provide a comparison value", message: nil, preferredStyle: .alert)
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
+        // Create an object to observe if text field input is empty.
+        var textFieldObserver: NSObjectProtocol!
+        let doneAction = UIAlertAction(title: "Done", style: .default) { _ in
+            let textField = alertController.textFields?.first
+            // Remove the empty string observer when done button is no longer in use.
+            NotificationCenter.default.removeObserver(
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift:182]: The observer need to be removed when cancel is tapped.

**Comment by yo1995** [comment]
> Thanks for your help @yo1995, I think it turned out great! I made some slight adjustments. Please let me know what you think and we can merge this to the original branch.

Sounds good! I fixed the merge conflict by merging v.next into this branch, so the changed files history looks a bit scary. 
I took a look and the improvements looks great. Feel free to give an approval. 🎉

**Comment by yo1995** [comment]
Oops sorry I missed one thing... Hold off for a second.

**Comment by yo1995** [comment]
> Oops sorry I missed one thing... Hold off for a second.

Should be good now! Case insensitiveness file system sometimes has issue with Git... 😅

**Pull request merged**